### PR TITLE
Create ckb-error and use it as the global system error type

### PR DESCRIPTION
**Pull request by keroro520**
* Create ckb-error and use it as the global system error type
* This PR is an update of #1344 based on https://github.com/nervosnetwork/ckb/pull/1344#issuecomment-520107924

* There are several error kinds, and the corresponding error types:

  * OutPoint: `util/types/src/core/error.rs`
  * Transaction: `verification/src/error.rs`
  * Script: `script/src/error.rs`
  * Header: `verification/src/error.rs`
  * Block: `verification/src/error.rs`
  * Internal: `error/src/internal.rs`
  * Dao: `util/dao/utils/src/error.rs`
  * Spec: `spec/src/error.rs`

This PR does not cover the error types in ckb-miner, ckb-rpc and ckb-sync. Perhaps do it later.

----

NOTE:

- I will remove the `NOTE: the original name is *` lines later. These lines are used to help review.

### CODE REVIEW

**Comment by keroro520** [comment]
@u2 Thanks for your careful review. I have updated  according to the above suggestions, in the new commitment https://github.com/nervosnetwork/ckb/commit/029e9025779c86348ea890d522c26aeb1c3d9ed6

* Rename some fields
* Distinguish Pow and Epoch errors
* Explicitly list the malformed errors

File: verification/src/error.rs [hunk 1]
```diff
@@ -96,119 +173,213 @@ pub enum CellbaseError {
...
         match self {
-            CapacityOverflow | DuplicateDeps | Empty | OutputsSumOverflow | InvalidScript
-            | ScriptFailure(_) | InvalidSignature | InvalidSince => true,
-            _ => false,
+            TransactionError::OutputOverflowCapacity
+            | TransactionError::DuplicatedDeps
+            | TransactionError::MissingInputsOrOutputs
+            | TransactionError::OccupiedOverflowCapacity
+            | TransactionError::InvalidSinceFormat
+            | TransactionError::TooLargeSize
```
Comment (u2) [inline verification/src/error.rs]: `OccupiedOverflowCapacity`, `UnmatchedOutputsDataLength`, `UnmatchedOutputsDataHashes` are `is_malformed_tx`, before they are not. But I think they should be.

But where is the `InvalidSignature` and `CapacityOverflow`?

File: error/src/internal.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    /// Database exception
+    // NOTE: the original name is ckb_db::Error::DBError(String)
+    Database,
+
+    /// VM internal error
+    VM,
+
+    /// The transaction already exist in pool
+    DuplicatedPoolTransaction,
+
```
Comment (zhangsoledad) [inline error/src/internal.rs]: IMO, the subject comes before adjective will be better. such as 
CapacityArithmeticOverflow,
TransactionPoolFull,
PoolTransactionDuplicated...

File: script/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#[derive(Fail, Debug, PartialEq, Eq, Clone)]
+pub enum ScriptError {
+    /// The field code_hash in script is invalid
+    #[fail(display = "InvalidCodeHash")]
+    InvalidCodeHash,
+
+    /// The script consumes too much cycles
+    // NOTE: the original name is ExceededMaximumCycles,
+    #[fail(display = "TooMuchCycles")]
+    TooMuchCycles,
```
Comment (zhangsoledad) [inline script/src/error.rs]: `TooMuchCycles` is really a bad name,  How much is too much?
the original name is much better, why replace?

**Comment by zhangsoledad** [comment]
I oppose all TooMuch TooMany-series rename

File: miner/src/block_assembler.rs [hunk 4]
```diff
@@ -465,7 +465,7 @@ impl BlockAssembler {
         let occupied_capacity = output.occupied_capacity(Capacity::bytes(data.len())?)?;
 
         if reward < occupied_capacity {
-            return Err(TransactionError::InsufficientCellCapacity.into());
+            Err(TransactionError::OccupiedOverflowCapacity)?;
```
Comment (zhangsoledad) [inline miner/src/block_assembler.rs]: OccupiedOverflowCapacity is weird.

File: error/src/internal.rs (hunk 2 above)
Comment (u2) [reply error/src/internal.rs]: We have discussed in the preview PR, but no conclusion.

**Comment by keroro520** [comment]
@zhangsoledad @u2 @quake 

I would suggest changing the error names and descriptions back to the original. And then open another issue to discuss it. What are your opinions?

File: verification/src/error.rs (hunk 1 above)
Comment (keroro520) [reply verification/src/error.rs]: `InvalidSignature` has been never used.

`CapacityOverflow` added.

**Comment by u2** [comment]
> @zhangsoledad @u2 @quake
> 
> I would suggest changing the error names and descriptions back to the original. And then open another issue to discuss it. What are your opinions?

Agree. You can create some small PRs for it, it's better for review also.

**Comment by keroro520** [comment]
> > @zhangsoledad @u2 @quake
> > I would suggest changing the error names and descriptions back to the original. And then open another issue to discuss it. What are your opinions?
> 
> Agree. You can create some small PRs for it, it's better for review also.

Updated: Change the error names back to the original.

Please help review again. Thanks~

**Comment by zhangsoledad** [comment]
> @zhangsoledad @u2 @quake
> 
> I would suggest changing the error names and descriptions back to the original. And then open another issue to discuss it. What are your opinions?

yeah, that's good!

File: sync/src/types.rs [hunk 5]
```diff
@@ -1208,12 +1207,12 @@ impl SyncSharedState {
...
+            self.peers()
+                .set_last_common_header(peer, block.header().clone());
         }
 
-        self.remove_header_view(&block.as_ref().hash());
-        self.peers()
-            .set_last_common_header(peer, block.header().clone());
-        ret
+        Ok(ret?)
     }
```
Comment (zhangsoledad) [inline sync/src/types.rs]: Oops :eyes:

Comment (keroro520) [reply sync/src/types.rs]: I write `Ok(ret?)` just for convenience to convert `Result<(), ckb_error::Error>` to `Result<(), FailureError>`.
If this is a bad style, I would like to change like before, maybe more explicitly.

File: verification/src/tests/uncle_verifier.rs [hunk 6]
```diff
@@ -628,6 +613,6 @@ fn test_uncle_with_uncle_descendant() {
         let dummy_context = dummy_context(&shared);
         let uncle_verifier_context = UncleVerifierContext::new(&dummy_context, &epoch18);
         let verifier = UnclesVerifier::new(uncle_verifier_context, &new_block);
-        assert_eq!(verifier.verify(), Ok(()));
+        assert!(verifier.verify().is_ok());
```
Comment (zhangsoledad) [inline verification/src/tests/uncle_verifier.rs]: I think `assert_eq!(verifier.verify(), Ok(()))` is more friendly, 
when assert failed, will got 
left: {error_kind} right: `Ok(())`

But `assert!(verifier.verify().is_ok())` only got 
left: false right: ture

Just comment here, not required, we shall improve it later for test friendly, someday.
`assert_eq!(ret, Ok(()), "testing sth got unexpected error {}", ret)`

File: sync/src/types.rs (hunk 5 above)
Comment (zhangsoledad) [reply sync/src/types.rs]: I don't mean that. the original code makes me little confused.

```
if is_err {
   return 

} else {
   ~~~~;
}
~~~~;
~~~~;
```

Comment (zhangsoledad) [reply sync/src/types.rs]: I don't mean that. the original code makes me little confused.

```
if is_err {
   return err;
} else {
  ...
}
...
why else
```

File: shared/src/tx_pool_ext.rs [hunk 7]
```diff
@@ -212,57 +208,71 @@ impl TxPool {
...
+                            // just be discarded.
+                            //
+                            // To avoid mis-discarding error types added in the future, please don't
+                            // use `Some(_)` as the match arm.
+                            //
+                            // OutOfOrder should only appear in BlockCellProvider
+                            OutPointError::InvalidHeaderDep(_)
+                            | OutPointError::InvalidDepGroup(_)
+                            | OutPointError::Empty(_)
+                            | OutPointError::UnspecifiedInputCell(_)
```
Comment (u2) [inline shared/src/tx_pool_ext.rs]: When does it raise `UnspecifiedInputCell` and `UnknownHeader`?

Comment (keroro520) [reply shared/src/tx_pool_ext.rs]: Oh, I didn't notice that they were removed from https://github.com/nervosnetwork/ckb/pull/1356

Thanks! I will remove the two errors.

File: verification/src/tests/uncle_verifier.rs (hunk 6 above)
Comment (keroro520) [reply verification/src/tests/uncle_verifier.rs]: Yes, I agree.

I just couldn't find a better way to fight against `rustc` about the compiling errors 😞 :

```
error[E0369]: binary operation `==` cannot be applied to type `std::result::Result<core::cell::ResolvedTransaction<'_>, ckb_error::Error>`
   --> util/types/src/core/cell.rs:707:9
    |
707 |         assert_eq!(result, Ok(()));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |         |
    |         std::result::Result<core::cell::ResolvedTransaction<'_>, ckb_error::Error>
    |         std::result::Result<(), _>
    |
    = note: an implementation of `std::cmp::PartialEq` might be missing for `std::result::Result<core::cell::ResolvedTransaction<'_>, ckb_error::Error>`
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to previous error
```

**Comment by u2** [comment]
bors r=quake,u2,zhangsoledad

**Comment by u2** [comment]
bors r=quake,u2,zhangsoledad

**Comment by zhangsoledad** [comment]
bors r=quake,u2,zhangsoledad

**Pull request merged**